src/main/java/com/epam/esm/dao/impl/HibernateGiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,145 @@
+package com.epam.esm.dao.impl;
+
+import com.epam.esm.dao.GiftCertificateDAO;
+import com.epam.esm.dao.request.CertificateRequestBody;
+import com.epam.esm.dao.service.PersistenceService;
+import com.epam.esm.dao.sort.SortBy;
+import com.epam.esm.dao.sort.SortType;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
+import com.epam.esm.service.exception.ErrorCodeEnum;
+import com.epam.esm.service.exception.ServiceException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.stereotype.Repository;
+
+import javax.annotation.PostConstruct;
+import javax.persistence.*;
+import javax.persistence.criteria.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+@Repository
+@EntityScan(basePackages = "com.epam.esm.model")
+public class HibernateGiftCertificateDaoImpl implements GiftCertificateDAO {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    private final PersistenceService<GiftCertificate> persistenceService;
+
+    private static final String GET_CERTIFICATE_BY_NAME = "SELECT g FROM GiftCertificate g WHERE g.name=:name";
+    private static final String GET_CERTIFICATE_COUNT = "SELECT count(g.id) FROM GiftCertificate g ";
+
+    @Autowired
+    public HibernateGiftCertificateDaoImpl(PersistenceService<GiftCertificate> persistenceService) {
+        this.persistenceService = persistenceService;
+    }
+
+    @PostConstruct
+    private void init() {
+        persistenceService.setType(GiftCertificate.class);
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificate(String name) throws NoResultException {
+        return persistenceService.getModelByName(GET_CERTIFICATE_BY_NAME, name);
+    }
+
+    @Override
+    public GiftCertificate getGiftCertificate(int id) {
+        return persistenceService.getModelById(id);
+    }
+
+    @Override
+    public List<GiftCertificate> getGiftCertificatesByRequestBody(
+            CertificateRequestBody requestBody, int page, int size) throws ServiceException {
+        if (requestBody == null) {
+            requestBody = CertificateRequestBody.getDefaultCertificateRequestBody();
+        }
+        if (!requestBody.getSortBy().equals(SortBy.NAME) && !requestBody.getSortBy().equals(SortBy.DATE)) {
```

Comment:
Because its the only sorting options available for certificate, and SortBy contains not only name and date but also price which isn't cost in certificate